### PR TITLE
Context Menu cleanup

### PR DESCRIPTION
**Pull request by stellar-aria**
This is just some general housekeeping for the `ContextMenu` classes that I felt should get done before/as part of working on #121. (Primarily for the runtime string-selection work I'm currently tackling there).

There was a small discussion on Discord Voice today about eventually working towards merging context_menu and menu_item (as context menus are just a specialization of menu that can be explicitly displayed in a popup, and that should be possible of all menus, etc).

Namespaces are currently following directory structure.

This should get checked on an OLED (explicitly for context menu titles and option listings) as I don't have an OLED and technically changed how the behavior works slightly (instance variable to method).

### CODE REVIEW

File: src/deluge/gui/context_menu/audio_input_selector.h [hunk 1]
```diff
@@ -21,15 +21,28 @@
...
+	bool canSeeViewUnderneath() override { return true; }
 
 	AudioOutput* audioOutput;
+
+	// Title
+	char const* getTitle() override;
+
+	// Options
+	size_t getNumOptions() override;
+	const char** getOptions() override;
```
Comment (bobtwinkles) [inline src/deluge/gui/context_menu/audio_input_selector.h:43]: Can we C++-ify this somehow? Having 2 separate functions sounds like a great trap for young players.

**Review by bobtwinkles — COMMENTED** [review]
Mostly LGTM functionally, thanks for chipping away in the refactoring mines a bit more. Will test on hardware tomorrow (if noone else gets to it first).

File: src/deluge/gui/context_menu/audio_input_selector.h [hunk 2]
```diff
@@ -21,15 +21,28 @@
 
 class AudioOutput;
 
+namespace deluge::gui::context_menu {
```
[the remark this answers is not in the document]
Comment (stellar-aria) [reply src/deluge/gui/context_menu/audio_input_selector.h:24]: GCC 12 defaults to c++17, and I'd honestly prefer to commit to c++20 if at all possible for things like concepts and consteval. I was planning on waiting for that request until we had e2 calling dbt, but maybe it's better to make it now?

File: src/deluge/gui/context_menu/sample_browser/kit.cpp [hunk 3]
```diff
@@ -15,22 +15,22 @@
...
-ContextMenuSampleBrowserKit contextMenuFileBrowserKit{};
+namespace deluge::gui::context_menu::sample_browser {
+Kit kit{};
 
-ContextMenuSampleBrowserKit::ContextMenuSampleBrowserKit() {
-#if HAVE_OLED
-	title = "Sample(s)";
-#endif
+char const* Kit::getTitle() {
+	static char const* title = "Sample(s)";
```
[the remark this answers is not in the document]
Comment (stellar-aria) [reply src/deluge/gui/context_menu/sample_browser/kit.cpp:29]: Yes, exactly. Titles shouldn't be a property of an object but a request they respond to anyways, following proper encapsulation.

File: src/deluge/gui/context_menu/audio_input_selector.h (hunk 1 above)
Comment (stellar-aria) [reply src/deluge/gui/context_menu/audio_input_selector.h:43]: The best way to do that would probably be returning a std::pair of the `char**`, and a `size_t` that can then be destructured in the caller...

File: src/deluge/gui/context_menu/audio_input_selector.h [hunk 4]
```diff
@@ -21,15 +21,28 @@
...
+	bool getGreyoutRowsAndCols(uint32_t* cols, uint32_t* rows) override;
+	void selectEncoderAction(int8_t offset) override;
 	bool setupAndCheckAvailability();
-	bool canSeeViewUnderneath() { return true; }
+	bool canSeeViewUnderneath() override { return true; }
 
 	AudioOutput* audioOutput;
+
+	// Title
+	char const* getTitle() override;
```
[the remark this answers is not in the document]
Comment (stellar-aria) [reply src/deluge/gui/context_menu/audio_input_selector.h:39]: They were honestly more just internal scribble-notes. I'll turn them into triple-slash comments to make that clear

**Comment by stellar-aria** [comment]
> I tested this on my OLED and it looks and behaves as expected, except for one issue: with the audio source selector menu for an audio clip, the initial value is OFF (first item), but the initial scroll position is set to the second item, i e you cannot see the selected OFF value. once you start move around the menu it looks like expected tho.

I miiiiight have just fixed that on accident? Is the problem still there with this latest commit?

**Comment by bfredl** [comment]
> I miiiiight have just fixed that on accident? Is the problem still there with this latest commit?

I tested again with the latest commit and it still seems to be a problem.

**Pull request merged**